### PR TITLE
[PyTorch] Normalization ops

### PR DESCRIPTION
**Pull request by timmoon10**
# Description

This PR extends the operation-based API (see https://github.com/NVIDIA/TransformerEngine/pull/707) with LayerNorm, RMSNorm, and FP8 cast operations.

Compare with the existing module-based API:
```python
# Module-based API
module1 = te.LayerNormLinear(...)

# Operation-based API
module2 = te.ops.Sequential(
    te.ops.LayerNorm(...),
    te.ops.Linear(...),
)
```

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refractor

## Changes

Please list the changes introduced in this PR:

- LayerNorm operation
- FP8 cast operation
- RMSNorm operation

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: transformer_engine/pytorch/module/layernorm.py [hunk 1]
```diff
@@ -3,156 +3,81 @@
...
-            torch.empty(
-                hidden_size,
-                device=device,
-                dtype=params_dtype,
-            )
-        )
-        self.sequence_parallel = sequence_parallel
 
-        self.reset_parameters(defer_init=(device == "meta"))
+        # Handle deprecated options
```
Comment (ptrendx) [inline transformer_engine/pytorch/module/layernorm.py:63]: This seems a little backwards. if `dtype` is supposed to be the new argument name, then why is it in `kwargs`? Both `params_dtype` and `dtype` should be regular parameters, there should be a deprecation warning when somebody uses `params_dtype` and also the check for duplicate assignment like the one you have here.
Also, similar treatment should be done for hidden_size and sequence_parallel (especially the last one seems to be just gone completely so there should be some explanation that it was unused before or something?)

File: transformer_engine/pytorch/module/layernorm.py [hunk 2]
```diff
@@ -3,156 +3,81 @@
...
-          The device on which the parameters of the model will allocated. It is the user's
-          responsibility to ensure all parameters are moved to the GPU before running the
-          forward pass.
+        If `True`, the :math:`\gamma` parameter is initialized to zero
+        and the calculation changes to
+
+            .. math::
+                y = \frac{x - \mathrm{E}[x]}{\sqrt{\mathrm{Var}[x] + \varepsilon}} * (1 + \gamma) + \beta
+
+    sm_margin: int, default = 0
```
Comment (ptrendx) [inline transformer_engine/pytorch/module/layernorm.py:47]: Maybe it should be tuple instead to handle FWD/BWD differently?

Comment (ptrendx) [reply transformer_engine/pytorch/module/layernorm.py:47]: The original had 3 things - fwd/bwd/inference

File: transformer_engine/pytorch/module/layernorm.py [hunk 3]
```diff
@@ -162,51 +87,61 @@ def reset_layer_norm_parameters(self) -> None:
...
-            self.zero_centered_gamma,
-            torch.is_grad_enabled(),
-            self.activation_dtype,
-        )
 
-        return fwd_fn(*args)
+        # Check whether to defer init (deprecated)
+        if defer_init is not None:
+            warnings.warn(
+                'reset_parameters kwarg is deprecated. Set device to "meta" instead.',
```
Comment (ptrendx) [inline transformer_engine/pytorch/module/layernorm.py:98]:
```suggestion
                'defer_init argument to reset_parameters function is deprecated. Set device to "meta" instead.',
```

File: transformer_engine/pytorch/module/layernorm.py [hunk 4]
```diff
@@ -162,51 +87,61 @@ def reset_layer_norm_parameters(self) -> None:
...
+                stacklevel=2,
+            )
+            if defer_init:
+                return
+
+        # Reset parameters
+        super().reset_parameters()
+
+        # Set flag for sequence parallelism (deprecated)
+        if getattr(self, "sequence_parallel", None) is not None:
```
Comment (ptrendx) [inline transformer_engine/pytorch/module/layernorm.py:109]: So why is sequence_parallel option deprecated then? I believe Megatron is using those to guide their logic for the optimizer. I know it is not great, but we should not break them.

File: transformer_engine/pytorch/module/rmsnorm.py [hunk 5]
```diff
@@ -3,207 +3,143 @@
...
+    r"""Root Mean Square Layer Normalization
 
-    where
+    Applies Root Mean Square Layer Normalization over a mini-batch of
+    inputs as described in the paper
+    `Root Mean Square Layer Normalization <https://arxiv.org/abs/1910.07467>`__
 
     .. math::
-        RMS_\varepsilon(x) = \sqrt{\frac{1}{n}\sum_{i=0}^nx_i^2 + \varepsilon}
+        y = \frac{x}{\sqrt{\mathrm{Var}[x] + \varepsilon}} * \gamma
```
Comment (ptrendx) [inline transformer_engine/pytorch/module/rmsnorm.py:24]: Is that true? Var would subtract the mean, no? And RMSNorm does not do that.

File: transformer_engine/pytorch/module/rmsnorm.py [hunk 6]
```diff
@@ -3,207 +3,143 @@
...
-            torch.empty(
-                hidden_size,
-                device=device,
-                dtype=params_dtype,
-            )
-        )
-        self.sequence_parallel = sequence_parallel
 
-        self.reset_parameters(defer_init=(device == "meta"))
+        # Handle deprecated options
```
Comment (ptrendx) [inline transformer_engine/pytorch/module/rmsnorm.py:62]: Same issue as in LN.

File: transformer_engine/pytorch/module/rmsnorm.py [hunk 7]
```diff
@@ -3,207 +3,143 @@
...
+        # Flag for sequence parallelism (deprecated)
+        self.sequence_parallel: Optional[bool] = sequence_parallel
 
     def reset_rms_norm_parameters(self) -> None:
-        """Init RMSNorm params"""
+        """Deprecated"""
         warnings.warn(
-            "This method is deprecated and will be removed in an upcoming release. "
-            "Update your code to use RMSNorm.reset_parameters() instead.",
+            "This method will be deprecated in an upcoming release. "
```
Comment (ptrendx) [inline transformer_engine/pytorch/module/rmsnorm.py:84]: If you issue this warning, then this function is already deprecated :-). It should say that it will be removed in an upcoming release.

File: transformer_engine/pytorch/module/rmsnorm.py [hunk 8]
```diff
@@ -3,207 +3,143 @@
...
+            "Update your code to use LayerNorm.reset_parameters() instead.",
```
Comment (ptrendx) [inline transformer_engine/pytorch/module/rmsnorm.py:85]: Why LayerNorm?

File: transformer_engine/pytorch/module/rmsnorm.py (hunk 7 above)
Comment (ptrendx) [reply transformer_engine/pytorch/module/rmsnorm.py:84]: So basically what was there before was alright.

File: transformer_engine/pytorch/module/rmsnorm.py [hunk 9]
```diff
@@ -3,207 +3,143 @@
...
-            self.inf_rmsnorm_sm_margin,
-            self.zero_centered_gamma,
-            torch.is_grad_enabled(),
-            self.activation_dtype,
-        )
+        self.reset_parameters()
 
-        return fwd_fn(*args)
+    def reset_parameters(self, defer_init: Optional[bool] = None) -> None:
+        """Init LayerNorm parameters"""
```
Comment (ptrendx) [inline transformer_engine/pytorch/module/rmsnorm.py:92]: RMSNorm

File: transformer_engine/pytorch/module/rmsnorm.py [hunk 10]
```diff
@@ -3,207 +3,143 @@
...
+
+        # Check whether to defer init (deprecated)
+        if defer_init is not None:
+            warnings.warn(
+                'reset_parameters kwarg is deprecated. Set device to "meta" instead.',
```
Comment (ptrendx) [inline transformer_engine/pytorch/module/rmsnorm.py:97]: Same as in LN

File: transformer_engine/pytorch/ops/basic/cast_float8.py [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+
+import torch
+
+from ...fp8 import FP8GlobalStateManager, get_fp8_te_dtype
+from ...tensor import Float8Tensor
+from ..op import BasicOperation, OperationContext
+from .._common import is_float8_tensor
+
+
+class CastFloat8(BasicOperation):
```
Comment (ptrendx) [inline transformer_engine/pytorch/ops/basic/cast_float8.py:18]: I believe you said that it is mostly an utility op for tests, right? We should probably mention that in this documentaiton.

Comment (ptrendx) [reply transformer_engine/pytorch/ops/basic/cast_float8.py:18]: Also maybe we should consider generalizing it a bit with a name that is not specific to FP8? (Like just Quantize)?

File: transformer_engine/pytorch/module/rmsnorm.py (hunk 5 above)
Comment (timmoon10) [reply transformer_engine/pytorch/module/rmsnorm.py:24]: Good catch
```suggestion
        y = \frac{x}{\text{RMS}_\varepsilon(x)} * \gamma

    where

    .. math::
        \text{RMS}_\varepsilon(x) = \sqrt{\frac{1}{n}\sum_{i=0}^n x_i^2 + \varepsilon}
```

File: transformer_engine/pytorch/module/rmsnorm.py (hunk 7 above)
Comment (timmoon10) [reply transformer_engine/pytorch/module/rmsnorm.py:84]:
```suggestion
            "This method is deprecated and will be removed in an upcoming release. "
```

File: transformer_engine/pytorch/module/rmsnorm.py (hunk 8 above)
Comment (timmoon10) [reply transformer_engine/pytorch/module/rmsnorm.py:85]:
```suggestion
            "Update your code to use RMSNorm.reset_parameters() instead.",
```

File: transformer_engine/pytorch/module/rmsnorm.py (hunk 9 above)
Comment (timmoon10) [reply transformer_engine/pytorch/module/rmsnorm.py:92]:
```suggestion
        """Init RMSNorm parameters"""
```

File: transformer_engine/pytorch/module/rmsnorm.py (hunk 10 above)
Comment (timmoon10) [reply transformer_engine/pytorch/module/rmsnorm.py:97]:
```suggestion
                'defer_init argument to reset_parameters function is deprecated. Set device to "meta" instead.',
```

File: transformer_engine/pytorch/module/layernorm.py (hunk 1 above)
Comment (timmoon10) [reply transformer_engine/pytorch/module/layernorm.py:63]: My thinking is that we should forward kwargs directly to `te.ops.LayerNorm` as much as possible so that we only have to change the API in one place if we ever make changes in the future. We include the deprecated options as explicit kwargs since they are specific to the module.

File: transformer_engine/pytorch/ops/basic/cast_float8.py (hunk 11 above)
Comment (timmoon10) [reply transformer_engine/pytorch/ops/basic/cast_float8.py:18]: It could be a helpful op for users as well. For example, if a user wants to have discrete layers for design reasons but still wants to fuse some operations with FP8 casts:
```python
act = te.ops.Sequential(te.ops.GeLU(), te.ops.Quantize())
linear = te.ops.Sequential(te.ops.Linear())
y = act(x)
z = linear(y)
```

File: transformer_engine/pytorch/module/layernorm.py (hunk 2 above)
Comment (timmoon10) [reply transformer_engine/pytorch/module/layernorm.py]: You can now pass in either an `int` or a `dict`.

File: transformer_engine/pytorch/module/layernorm.py (hunk 4 above)
Comment (timmoon10) [reply transformer_engine/pytorch/module/layernorm.py:119]: I've changed the comment to read:
```
# Set flag for sequence parallelism (custom Megatron-LM integration)
```

**Comment by timmoon10** [comment]
Merging with approval from @ptrendx and @ksivaman.

**Comment by binxuan** [comment]
Hi one question regarding this new layernorm implementation, my understanding is that this new implementation can support multi-dimension layernorm weight while previous implemenation can only support one-dimension weight. 

So I have N different 1-D tensors and previously I have to initiate N different layernorm and apply them separately. But with this new implementation, then we can apply one 2-dimensional layernorm to N stacked different tensors with shape (N, dim)? is my understanding correct?

**Comment by timmoon10** [comment]
@binxuan This implementation matches [`torch.nn.LayerNorm`](https://pytorch.org/docs/stable/generated/torch.nn.LayerNorm.html):
```python
x_2d = x.reshape(-1, prod(normalized_shape))
y_2d = torch.nn.functional.layer_norm(x_2d, ...)
y = y_2d.reshape(x.size())
```
